### PR TITLE
Moving the client.keys synchronization mechanism with wazuh-db from the synchronization module to authd in the master nodes

### PR DESCRIPTION
**Pull request by DProvinciani**
|Related issue|
|---|
| https://github.com/wazuh/wazuh/issues/11403 |

## Description

This pull request implements the next set of changes:

- The addition of the agents to the `global.db` database now is performed by the writer thread in `authd` in the master node.
- In the synchronization module, the `client.keys` synchronization with `wazuh-db` now is performed only in worker nodes.
- The synchronization mechanism of `client.keys` with `wazuh-db` used by the synchronization module was moved to a common source code file.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [x] Package upgrade

- Memory tests for Linux
  - [x] Scan-build report

### CODE REVIEW

**Comment by DProvinciani** [comment]
### Performance analysis

The next script was created to perform a time measure of the time spent to add 100 agents in version 4.2.5 and the development branch. No performance degradations were detected in the measurements after running the script three times on each version. In fact, sometimes we saw a time reduction as shown in the snippets.

```bash
#!/bin/bash
TOKEN=$(curl -u wazuh:wazuh -k -X GET "https://localhost:55000/security/user/authenticate?raw=true")

start_time=$(date +%s.%3N)
for i in {1..100}
do
   curl -k -X POST "https://localhost:55000/agents" -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json" -d "{\"name\":\"agent_$i\"}"
   echo ""
done
end_time=$(date +%s.%3N)

elapsed=$(echo "scale=3; $end_time - $start_time" | bc)
echo "Elapsed time: $elapsed"
```

## Time measure in `4.2.5`
perf-add-agent-425

## Time measure in the development branch
perf-add-agent-440

**Comment by DProvinciani** [comment]
### `scan-build` results
scan-build-ok

File: src/wazuh_modules/wm_database.c [hunk 1]
```diff
@@ -565,7 +565,6 @@ int wm_sync_file(const char *dirname, const char *fname) {
         }
 
         free(name);
-        result = wm_sync_agent_group(id_agent, fname);
```
Comment (pereyra-m) [inline src/wazuh_modules/wm_database.c:568]: I think that the whole `case WDB_GROUPS:` section can be removed, now that the groups files don't exist any more.

There are another calls related, like `wm_scan_directory(GROUPS_DIR);` that should be also considered for removal.

File: src/os_auth/main-server.c [hunk 2]
```diff
@@ -837,12 +837,21 @@ void* run_writer(__attribute__((unused)) void *arg) {
...
 
         for (cur = copy_insert; cur; cur = next) {
             next = cur->next;
 
             mdebug1("[Writer] Performing insert([%s] %s).", cur->id, cur->name);
+
+            gettime(&t0);
+            if (wdb_insert_agent(atoi(cur->id), cur->name, NULL, cur->ip, cur->raw_key, cur->group, 1, &wdb_sock)) {
+                // The agent already exists, update group only.
+                mdebug2("The agent %s '%s' already exist in the database.", cur->id, cur->name);
```
Comment (pereyra-m) [inline src/os_auth/main-server.c:850]: Typo: `exist`

**Review by pereyra-m — CHANGES_REQUESTED** [review]
### Review

Only some minor comments, great job!
All the `os_free()` added are a good contribution.

### Review aspects:

- Code style: improvements suggested
- Typo: correct writing of log messages
- Implementation goals: the PR fulfills the requirements 
- Risks: synchronization conflicts, race conditions and memory leaks were not detected

File: src/os_auth/main-server.c [hunk 3]
```diff
@@ -837,12 +837,21 @@ void* run_writer(__attribute__((unused)) void *arg) {
...
+        os_free(copy_keys);
 
         for (cur = copy_insert; cur; cur = next) {
             next = cur->next;
 
             mdebug1("[Writer] Performing insert([%s] %s).", cur->id, cur->name);
+
+            gettime(&t0);
+            if (wdb_insert_agent(atoi(cur->id), cur->name, NULL, cur->ip, cur->raw_key, cur->group, 1, &wdb_sock)) {
+                // The agent already exists, update group only.
```
[the remark this answers is not in the document]
Comment (DProvinciani) [reply src/os_auth/main-server.c:849]: Fixed

File: src/os_auth/auth.c [hunk 4]
```diff
@@ -38,10 +38,12 @@ void add_insert(const keyentry *entry,const char *group) {
     node->id = strdup(entry->id);
     node->name = strdup(entry->name);
     node->ip = strdup(entry->ip->ip);
-    node->group = NULL;
+    node->raw_key = strdup(entry->raw_key);
 
     if (group != NULL)
         node->group = strdup(group);
+    else
```
[the remark this answers is not in the document]
Comment (DProvinciani) [reply src/os_auth/auth.c:45]: Fixed

File: src/wazuh_modules/wm_database.c (hunk 1 above)
Comment (DProvinciani) [reply src/wazuh_modules/wm_database.c:568]: I only removed the code that you suggested. There is more code to be deprecated but that will be done in a separate issue.
DONE.

File: src/os_auth/main-server.c (hunk 2 above)
Comment (DProvinciani) [reply src/os_auth/main-server.c:850]: Fixed

**Review by pereyra-m — APPROVED** [review]
LGTM !

**Pull request merged**